src/mscorlib/src/System/String.Manipulation.cs

```diff
@@ -1231,20 +1233,20 @@ private String[] SplitInternal(String separator, String[] separators, Int32 coun
                 return new String[] { this };
             }
 
-            int[] sepList = new int[Length];
-            int[] lengthList;
+            Span<int> sepList = Length < StackallocStringLengthLimit ? stackalloc int[Length] : new int[Length];
+            Span<int> lengthList = singleSeparator 
```

Comment:
would it be cheaper to do ` Span<int> lengthList = new Span<int>()` here and put `Length < StackallocStringLengthLimit ? stackalloc int[Length] : new int[Length];` inside the `else` clause below? That saves a branch, but adds newing up a struct, so I don't know.

Reply:
The most expensive part in the current code is variable sized-stackalloc. Fixed-size stackalloc is much cheaper,..

Reply:
Curious why that is? Both cases zero the memory right, or does the compiler determine it doesn't need to?

Reply:
> Both cases zero the memory right

No. We have zero-init for stackalloc disabled in CoreLib.

Reply:
Along the lines of @danmosemsft's comment, why not just keep the existing structure of the code, which would avoid having to check `singleSeparator` twice? e.g.:

```c#
Span<int> sepList = Length < StackallocStringLengthLimit ? stackalloc int[Length] : new int[Length];
Span<int> lengthList;
int defaultLength;
int numReplaces;

if (singleSeparator)
{
    lengthList = default;
    defaultLength = separator.Length;
    numReplaces = MakeSeparatorList(separator, sepList);
}
else
{
    lengthList = Length < StackallocStringLengthLimit ? stackalloc int[Length] : new int[Length];
    defaultLength = 0;
    numReplaces = MakeSeparatorList(separators, sepList, lengthList);
}
```

Reply:
@justinvp it does not compile with ` A result of a stackalloc expression of type 'Span<int>' canno
t be used in this context because it may be exposed outside of the containing method`

Reply:
Try making it `Span<int> lengthList = stackalloc int[0];`, e.g.:

```c#
Span<int> sepList = Length < StackallocStringLengthLimit ? stackalloc int[Length] : new int[Length];
Span<int> lengthList = stackalloc int[0];
int defaultLength;
int numReplaces;

if (singleSeparator)
{
    defaultLength = separator.Length;
    numReplaces = MakeSeparatorList(separator, sepList);
}
else
{
    lengthList = Length < StackallocStringLengthLimit ? stackalloc int[Length] : new int[Length];
    defaultLength = 0;
    numReplaces = MakeSeparatorList(separators, sepList, lengthList);
}
```


Reference: https://github.com/dotnet/corefx/pull/25426#discussion_r152644870

> Try making this Span<int> temp = stackalloc int[0];
> 
> Otherwise the local is classified as returnable and later you cannot mix it with stackallocated ones
> 
> stackalloc int[0]; is basically a noop, but will have effect of marking the temp as not returnable.

Reply:
that works